compiler/src/iree/compiler/Codegen/SPIRV/SPIRVLinkExecutables.cpp

```diff
@@ -40,22 +60,110 @@ struct SPIRVLinkExecutablesPass final
     }
     sourceExecutableOps.resize(retainSize);
 
-    // Guess a module name, if needed, to make the output files readable.
-    std::string moduleName = guessModuleName(moduleOp, "spirv_module");
+    // Note that at runtime, for a particular executable, only one variant of it
+    // will be loaded. So, all variants of an executable are expected to provide
+    // the exact same set of entry points; this way we can guarantee no matter
+    // which variant is chosen, we have all entry points to call into. The same
+    // entry point in different variants may have different target requirements
+    // though.
+    //
+    // The input to the linking stage are a collection of executables, each may
+    // have multiple variants, but only ever provide one entry point. Together
+    // with the above restriction, we can link two executables if and only if
+    // their variants have the exact same set of target requirements. Under such
+    // circumstances, we can make sure for a particular target requirement
+    // (loaded as one variant during runtime), we can provide all entry points.
+
+    // Build a map from all variants' target requirements to their wrapping
+    // executable ops.
+    std::map<SmallVector<ExecutableTargetAttr, 0>,
+             SmallVector<IREE::HAL::ExecutableOp>>
+        executableBuckets;
+
+    SmallVector<ExecutableTargetAttr, 0> currentTargets;
+    for (int i = 0, e = sourceExecutableOps.size(); i < e; ++i) {
+      IREE::HAL::ExecutableOp executable = sourceExecutableOps[i];
```

Comment:
nit: ? I don't see the loop iterator used anywhere.
```suggestion
    for (auto executable : sourceExecutableOps) {
```